rare/widgets/image_widget.py

```diff
@@ -0,0 +1,142 @@
+
+    def setPixmap(self, pixmap: QPixmap) -> None:
+        if not pixmap.isNull():
+            self._pixmap = pixmap
+            self.paint_image = self.paint_image_cover
+            if not self._image_size:
+                self._transform = QTransform().scale(
+                    1 / pixmap.devicePixelRatioF(),
+                    1 / pixmap.devicePixelRatioF()
+                )
+            else:
+                self._transform = QTransform().scale(
+                    1 / pixmap.devicePixelRatioF() / self._image_size.divisor,
+                    1 / pixmap.devicePixelRatioF() / self._image_size.divisor
+                )
+            self.update()
+
+    def setFixedSize(self, a0: ImageSize.Preset) -> None:
+        self._squared_overlay = None
+        self._rounded_overlay = None
+        self._image_size = a0
+        self._smooth_transform = a0.smooth
+        super(ImageWidget, self).setFixedSize(a0.size)
+
+    def setBorder(self, border: Border):
+        if border == ImageWidget.Border.Rounded:
+            self.paint_overlay = self.paint_overlay_rounded
+        else:
+            self.paint_overlay = self.paint_overlay_squared
+        self.update()
+
+    def _generate_squared_overlay(self) -> OverlayPath:
+        if self._image_size is not None and self._squared_overlay is not None:
+            return self._squared_overlay
+        path = QPainterPath()
+        path.addRect(0, 0, self.width(), self.height())
+        border = 2
+        inner_path = QPainterPath()
+        inner_path.addRect(
+            QRectF(
+                border,
+                border,
+                self.width() - border * 2,
+                self.height() - border * 2,
+            )
+        )
+        gradient = QLinearGradient(0, 0, 0, self.height())
+        gradient.setColorAt(0.0, Qt.black)
+        gradient.setColorAt(1.0, Qt.transparent)
+        self._squared_overlay = path.subtracted(inner_path), gradient
+        return self._squared_overlay
+
+    def _generate_rounded_overlay(self) -> OverlayPath:
+        if self._image_size is not None and self._rounded_overlay is not None:
+            return self._rounded_overlay
+        # lk: the '-1' and '+1' are adjustments required for anti-aliasing
+        # lk: otherwise vertical lines would appear at the edges
+        path = QPainterPath()
+        path.addRect(-1, -1, self.width() + 2, self.height() + 2)
+        rounded_path = QPainterPath()
+        rounded_path.addRoundedRect(
+            QRectF(0, 0, self.width(), self.height()),
+            self.height() * 0.045,
+            self.height() * 0.045,
+        )
+        self._rounded_overlay = path.subtracted(rounded_path), QColor(0, 0, 0, 0)
+        return self._rounded_overlay
+
+    def paint_image_empty(self, painter: QPainter, a0: QPaintEvent) -> None:
+        # when image is not available, show a gray rectangle
+        painter.setOpacity(0.5 * self._opacity)
+        painter.fillRect(a0.rect(), Qt.darkGray)
```

Comment:
Didn't you add my "default" game cover from before for exactly that? Or is it supposed to be a first step when drawing? Either ways, clarify.

Reply:
In this case the availability has to do with the pixmap object itself, not the availability of the cover. This is the case when the image has not been downloaded or loaded from disk yet. It draws a gray rectangle in its place.